src/sentry/snuba/metrics/fields/base.py

```diff
@@ -147,15 +175,212 @@ def _build_conditional_aggregate_for_metric(self, entity):
+            f"SingularEntityDerivedMetric {self.metric_name} after the `entity` attribute "
+            f"is set through calling `get_entity` with projects attribute"
+        )
+
+    @classmethod
+    def __recursively_get_all_entities_in_derived_metric_dependency_tree(
+        cls, derived_metric_name: str, projects: Sequence[Project]
+    ) -> Set[MetricEntity]:
+        """
+        Method that gets the entity of a derived metric by traversing down its dependency tree
+        until it gets to the raw metrics (leaf nodes) then queries snuba to check which
+        entity/entities these raw constituent metrics belong to.
+        """
+        if derived_metric_name not in DERIVED_METRICS:
+            metric_type = get_single_metric_info(projects, derived_metric_name)["type"]
+            return {METRIC_TYPE_TO_ENTITY[metric_type].value}
+        entities = set()
+        derived_metric = DERIVED_METRICS[derived_metric_name]
+        for metric_name in derived_metric.metrics:
+            entities |= cls.__recursively_get_all_entities_in_derived_metric_dependency_tree(
+                metric_name, projects
+            )
+        return entities
+
+    def get_entity(
+        self, projects: Optional[Sequence[Project]] = None, **kwargs: Any
+    ) -> MetricEntity:
+        if self._entity:
+            return self._entity
+        if not projects:
+            self.__raise_entity_validation_exception("get_entity")
+        entities = self.__recursively_get_all_entities_in_derived_metric_dependency_tree(
+            derived_metric_name=self.metric_name, projects=projects
+        )
+        if len(entities) != 1 or entities == {None}:
+            raise DerivedMetricParseException(
+                f"Derived Metric {self.metric_name} cannot be calculated from a single entity"
+            )
+        self._entity = entities.pop()
+        return self._entity
+
+    @classmethod
+    def __recursively_generate_metric_ids(cls, derived_metric_name: str) -> Set[int]:
+        """
+        Method that traverses a derived metric dependency tree to return a set of the metric ids
+        of its constituent metrics
+        """
+        if derived_metric_name not in DERIVED_METRICS:
+            return set()
+        derived_metric = DERIVED_METRICS[derived_metric_name]
+        ids = set()
+        for metric_name in derived_metric.metrics:
+            if metric_name not in DERIVED_METRICS:
+                ids.add(resolve_weak(metric_name))
+            else:
+                ids |= cls.__recursively_generate_metric_ids(metric_name)
+        return ids
+
+    def generate_metric_ids(self) -> Set[int]:
+        if not self._entity:
+            self.__raise_entity_validation_exception("generate_metric_ids")
+        return self.__recursively_generate_metric_ids(derived_metric_name=self.metric_name)
+
+    @classmethod
+    def __recursively_generate_select_snql(cls, derived_metric_name: str) -> List[Function]:
```

Comment:
Extending the idea of having child metrics of type `MetricFieldBase` above, these recursive methods could probably share some code by defining a single recursive generator like

```python
def descendants(self):
   yield self
   for child in self.child_metrics:
       yield from child.descendants()
```

and then get whatever you need from each descendant.